### ISSUE TITLE
Assert.Equal failing for value-slot collections of different concrete types

### ISSUE THREAD

**Issue by FlorianHockmann** [xunit/xunit#2850]
One of our tests starts failing in version 2.6.2 and 2.6.3, but it still worked in 2.6.1 (and versions before that).
The test compares two collections which contain a Dictionary that contains collections as its values. One of these nested collections is an `object[]` whereas the other is a `List<object>`.

This is a minimal example to reproduce the problem:

```c#
var expected = new List<object> { new Dictionary<object, object> { { 1, new List<object> { 2 } } } };
var actual = new object[] { new Dictionary<object, object> { { 1, new object[] { 2 } } } };

Assert.Equal(expected, actual);
```

It fails with:

```
Assert.Equal() Failure: Exception thrown during comparison
                        ↓ (pos 0)
Expected: List<object> [[[1] = [2]]]
Actual:   object[]     [[[1] = [2]]]
                        ↑ (pos 0)

System.InvalidCastException
Unable to cast object of type 'System.Object[]' to type 'System.Collections.Generic.List`1[System.Object]'.
```

The `InvalidCastException` comes from the nested collection in the Dictionary, not the outer collection.

**Comment by bradwilson** [comment]
Bummer. That means the fix for #2821 broke this. Not sure what the fix is yet.

**Comment by bradwilson** [comment]
This also affects both keys and values for comparing `KeyValuePair<K,V>`.

**Comment by bradwilson** [comment]
Available in v2: `2.6.4-pre.5`
Available in v3: `0.1.1-pre.334`

https://xunit.net/docs/using-ci-builds

**Comment by FlorianHockmann** [comment]
Thanks a lot for the really fast response and fix! :smiley:
